### PR TITLE
make: add blob utility header

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR adds a utility makefile that allows arbitrary files to be made available as headers.
This is supposed to be used with, e.g., lua, jerryscript, micropython for scripting language source files.

### Testing procedure

Contains a test application. Check CI output.

Also, try changing the blobs and see if everything gets rebuilt correctly. Hammer this using ```make -j ...```.
Do the same with ```BUILD_IN_DOCKER=1```.

### Issues/PRs references

Simpler alternative to #11497.

### CODE REVIEW

**Comment by emmanuelsearch** [comment]
@smlng ping ;)

**Comment by kaspar030** [comment]
To provide context vs #11497, this PR unifies *current* blob handling (which is duplicated in examples/javascript, examples/lua_REPL and examples/lua_basic, and one more coming in #2968) to a shared makefile. It is still using xxd & sed, which #11497 tries to replace with a custom make-based implementation.

IMO, while I disagree with handling this using make itself, even should we agree on that, this PR provides a clean entry-point to do that. Thus the guts of #11497 could be hooked in easily by replacing the one xxd+sed line of this PR. But that should be a different discussion.

File: tests/blob/tests/01-run.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+import sys
+from testrunner import run
+
+
+def testfunc(child):
+    child.expect(u"Hello blob!")
```
Comment (aabadie) [inline tests/blob/tests/01-run.py]:
```suggestion
    child.expect_exact("Hello blob!")
```

File: tests/blob/Makefile [hunk 2]
```diff
@@ -0,0 +1,10 @@
+DEVELHELP ?= 0
+include ../Makefile.tests_common
+
+USEMODULE += fmt
+
+BLOBS += blobtest.txt
+
+TEST_ON_CI_WHITELIST += all
```
Comment (aabadie) [inline tests/blob/Makefile]: This is not needed anymore, right ?!

File: makefiles/blob.inc.mk [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#    #include "blob/foo.ext.h"
+#
+
+# use "blobs/blob" so the headers can be included as "blob/foo.h", but
+# we don't have to add $(BINDIR)/$(MODULE) to the include path.
+BLOB_HDR_DIR ?= $(BINDIR)/$(MODULE)/blobs/blob
+BLOB_H := $(BLOBS:%=$(BLOB_HDR_DIR)/%.h)
+
+ifneq (, $(BLOB_H))
+	# add $(BINDIR)/$(MODULE)/blobs to include path
```
Comment (aabadie) [inline makefiles/blob.inc.mk]: Looks like there's a tab instead of 2 spaces here (and the line below)

File: tests/blob/Makefile (hunk 2 above)
Comment (kaspar030) [reply tests/blob/Makefile]: yup, removed

File: makefiles/blob.inc.mk (hunk 3 above)
Comment (kaspar030) [reply makefiles/blob.inc.mk]: fixed

**Comment by kaspar030** [comment]
@aabadie I accidentally over-force-pushed and then re-applied your suggestion commit.

File: makefiles/application.inc.mk [hunk 4]
```diff
@@ -3,4 +3,6 @@ MODULE = $(APPLICATION_MODULE)
 DIRS += $(RIOTCPU)/$(CPU) $(RIOTBOARD)/$(BOARD)
 DIRS += $(RIOTBASE)/core $(RIOTBASE)/drivers $(RIOTBASE)/sys
 
+BLOBS = $(APPLICATION_BLOBS)
```
Comment (aabadie) [inline makefiles/application.inc.mk]: Why not move this line to `makefiles/blob.inc.mk` ? Since this file is included by Makefile.base, it will be more consistent to handle this in the related file. This is untested so maybe this doesn't work because I'm missing something.

Comment (kaspar030) [reply makefiles/application.inc.mk]: If that line would be in blob.inc.mk, it would be used in every recursive make call (e.g., for every module). We only want it to be used for the application module. All other modules can directly use BLOBS.

From the corresponding commit description:

```

@kaspar030
make: pass BLOBS to makefiles/application.inc.mk

For regular modules, adding files to BLOBS is sufficient to create the
corresponding headers.

Application modules are different, as they use a minimal makefile
(makefiles.application.inc.mk) to build, thus application level
variables are not available.

This commit makes Makefile.include pass BLOBS to the application
Makefile as APPLICATION_BLOBS, and application.inc.mk use that variable
as value for BLOBS.

The indirection is necessary so submakefiles (e.g., those visited by
DIRS) do not hard override BLOBS.
```

Comment (aabadie) [reply makefiles/application.inc.mk]: Is it possible to add a comment about that here instead of inside the commit message ?

**Comment by aabadie** [comment]
I was asked to have a look at this PR but I don't feel confident enough to give a valuable review on the Makefile changes. Especially because they touch `Makefile.base`. Maybe @cladmi or @jcarrano could be of more help here ?

I have more general comments btw:
From the commit message linked above:
> For regular modules, adding files to BLOBS is sufficient to create the
corresponding headers.

So this means that this PR allows to embed blob defined in modules. The problem is that the test application doesn't show this. It's only applied to application blobs.

Another aspect is that the current implementation only generates a header from a text file and include this to the build. Conceptually, this is not really a blob.

**Comment by jcarrano** [comment]
I'll summarize a discussion we had AFK with @emmanuelsearch. Sorry for the long comment, and I don't mean to bash this PR- **I feel that many of the ideas expressed here apply equally to other instances where we may be wanting to "abstract away" common functionality.**

All this would achieve is remove the need for the following snippet to exist in the application's makefile:

```make
JS_PATH = $(BINDIR)/js/$(MODULE)

# add directory of generated *.js.h files to include path
CFLAGS += -I$(JS_PATH)

# generate .js.h header files of .js files
JS = $(wildcard *.js)
JS_H = $(JS:%.js=$(JS_PATH)/%.js.h)

BUILDDEPS += $(JS_H) $(JS_PATH)/

include $(RIOTBASE)/Makefile.include

$(JS_PATH)/:
	$(Q)mkdir -p $@

$(JS_H): $(JS_PATH)/%.js.h: %.js | $(JS_PATH)/
	$(Q)xxd -i $< | sed 's/^unsigned/const unsigned/g' > $@
```

That was taken from the jerryscript package. Lua has something similar. Note that header files will end up in a js-specific path.

The propose module removes the need for this code, at the price of:

- Reduced flexibility.
  - User cannot have different paths for different file types.
  - The datatype (const unsigned) is fixed.
    - In fact, the whole output template is fixed.
- The code is now "officially" part of the RIOT build system, which means people will depend on in and each change is an API change.

The reduced flexibility means that any deviation from the assumption that this new module does would require reverting back to the above snippet.

Now, there are some very real problems which are not being solved by this or by anything currently in use. Many of these are actually limitations of our build system, unrelated to this PR.

- Making files into arrays is easy, but the interesting part is _indexing_ them (i.e. putting them into some sort of structure or table)
  - I have come to believe that this is best done in an application specific way.
- We still have to come up with a way to generate _just_ the required dependencies. 
  - `$(OBJC) $(OBJCXX): $(BLOB_H)` (or builddeps...) is overkill, but we currently know no better way
- What if I want to generate an object instead of a header.
  - Needed, for example, to properly integrate stuff like vendor supplied binaries/ bootloaders /etc.
  - Compiling a generated C file and  is tricky.
  - #11497 does that, but I did not insist too much on that because the rest of the complaints in this PR apply to it as well (my justification for that PR was that it is something tricky to get right).

In conclusion, I'm against too much convenience functions, since these have to be maintained too.

**Comment by kaspar030** [comment]
> * User cannot have different paths for different file types.

If there's a use-case for that, it can be added. (e.g., by properly dealing with subfolders so blobs don't have to be in the module's folder.)

> * The datatype (const unsigned) is fixed.

Did you encounter this use case before?
Anyhow, this would be easy to add, maybe using BLOBS_VAR in addition to BLOBS.

>   * In fact, the whole output template is fixed.
> I have come to believe that this is best done in an application specific way.

I agree.

>     * What if I want to generate an object instead of a header.

Well, linking an object is a different function than creating data that is accessible from C from a binary.

**Comment by kaspar030** [comment]
> If there's a use-case for that, it can be added.

Well, I did, now ```BLOBS += $(wildcard js/*.js)``` should work correctly.

**Comment by kaspar030** [comment]
> So this means that this PR allows to embed blob defined in modules. The problem is that the test application doesn't show this. It's only applied to application blobs.

- added a module using a blob

> Another aspect is that the current implementation only generates a header from a text file and include this to the build. Conceptually, this is not really a blob.

- added a binary blob

File: tests/blob/blob_module/Makefile [hunk 5]
```diff
@@ -0,0 +1,2 @@
+BLOBS += blobtest.bin
```
Comment (aabadie) [inline tests/blob/blob_module/Makefile]: This file is missing ;)

Comment (kaspar030) [reply tests/blob/blob_module/Makefile]: thanks! It was blanked out by .gitignore...

**Comment by kaspar030** [comment]
- added some possible testing methods

File: makefiles/blob.inc.mk [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+# Usage:
+#
+# Add this to an application or module Makefile:
+#
+#     BLOBS += foo.ext
+#
+# Then include in C file or header:
+#
+#    #include "blob/foo.ext.h"
+#
```
Comment (bergzand) [inline makefiles/blob.inc.mk]: Could you also include the generated variable names in the example here?
Something like:
```
#include "blob/foo.ext.h"

print((char *)foo_ext, foo_ext_len);
```

**Comment by bergzand** [comment]
On some Linux based operating systems (e.g. Archlinux, Gentoo, Solus) the `xxd` utility is supplied with `vim`, on others it is a stand-alone package. I don't think I would call it universally installed on systems. Does it make sense to add a test if it is installed/executable before attempting to generate the files?

File: makefiles/blob.inc.mk (hunk 6 above)
Comment (kaspar030) [reply makefiles/blob.inc.mk]: I've added some docs.

**Comment by kaspar030** [comment]
> Does it make sense to add a test if it is installed/executable before attempting to generate the files?

Did! unfortunately "check_cmd" was not available from Makefile.base, so I replicated it.

File: makefiles/application.inc.mk (hunk 4 above)
Comment (kaspar030) [reply makefiles/application.inc.mk]: Now it is here, is that sufficient?

**Comment by bergzand** [comment]
Is there somebody available to give this a test? I'm willing to ack this, but unfortunately I'm unable tot test this today. Feel free to dismiss my review if required.

**Comment by aabadie** [comment]
> Is there somebody available to give this a test?

Tested locally on native. The structure of directories containing the blobs reflect what's done in this PR. So it looks good for me.

**Comment by aabadie** [comment]
(and it works, of course)

**Comment by bergzand** [comment]
@kaspar030 Please squash!

**Comment by bergzand** [comment]
Is there any reason here to consider this 'high impact' and wait until after the hard freeze?

**Comment by kaspar030** [comment]
> Is there any reason here to consider this 'high impact' and wait until after the hard freeze?

I don't think so. "blob.inc.mk" is always included and it does enable SECONDEXPANSION, but that shouldn't cause any problem. It doesn't in the CI. Maybe @cladmi has an opinion?

**Comment by cladmi** [comment]
`SECONDEXPANSION` ? wow we never needed it before.

**Comment by kaspar030** [comment]
> `SECONDEXPANSION` ? wow we never needed it before.

Yup. It is convenient here to make a target depend on its own folder. We might want to adapt that to other SRCS...

**Comment by kaspar030** [comment]
> `SECONDEXPANSION` ? wow we never needed it before.

but do you see any issue with it?

**Comment by cladmi** [comment]
It means I need to do a proper review to give any opinion.

**Comment by kaspar030** [comment]
> It means I need to do a proper review to give any opinion.

[Here](http://make.mad-scientist.net/secondary-expansion/) is some information of possible side effects of enabling SECONDEXPANSION. The two things are

1. it breaks prerequisite names that have dollar signs in the name (which are escaped by using double dollar signs). Luckily we don't do that.

2. performance. Older make versions handled this inefficiently. Apparently newer makes have this optimized so it doesn't cause a performance hit.

IMO this is safe to use.

**Comment by emmanuelsearch** [comment]
@bergzand so is that an ACK for real now ? ;)

**Comment by emmanuelsearch** [comment]
and @cladmi are you actually planning to review?

**Comment by aabadie** [comment]
@kaspar030, do you think it's possible to achieve the same result without using SECONDEXPANSION ?
In the current state, all targets defined after are impacted, even the ones that comes after the include of `blob.inc.mk` in `Makefile.base`. This means for all those targets we'll have to be very careful not escaping them.
@cladmi, do you think you'll have time to do a proper review and also give your feedback soon ?

**Comment by bergzand** [comment]
> @bergzand so is that an ACK for real now ? ;)

Seeing the concerns raised by @cladmi, I'm no longer sure. I'm not confident enough in my makefile skills to merge this right before the hard freeze. Are there any issues with merging this after the hard freeze?

File: makefiles/blob.inc.mk [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+$(BLOB_H): $(BLOB_HDR_DIR)/.
+$(BLOB_H): $(BLOB_HDR_DIR)/%.h: % $(BLOBS) | $$(@D)/.
+	@command -v $(XXD) > /dev/null 2>&1 || ( \
+	  echo "error: xxd binary '$(XXD)' not found! (Maybe install vim package?)"; exit 1)
+
+	${Q}cd $(dir $<); xxd -i $(notdir $<) | sed 's/^unsigned/const unsigned/g'> $@
+
+# make C and C++ objects of this module depend on generated headers, so they
+# get re-build on changes to the blob files.
+$(OBJC) $(OBJCXX): $(BLOB_H)
```
Comment (cladmi) [inline makefiles/blob.inc.mk]: The comment is wrong, rebuilding would be done automatically from the `.d` generated files as the compilation is done with `-MD -MP`.
This also does not account for deleted files that the `.d` does.

This is only currently required to have the headers generated before compiling the C files, so with a `: |$(BLOB_H)` it would work too.

**Comment by cladmi** [comment]
It looks like the `SECONDEXPANSION` is only required to allow having subdirectories for headers inside `blob`. Is this a required feature?

If it really is, the `SECONDEXPANSION` is not really required here.
The `$$(@D)` handling is only needed if the targets list is unknown to have a generic handling or pattern rules.

Here as all targets are basically known and could be handled with a stupid handling that all targets depend on having created all the directories first:

```
BLOB_DIRECTORIES = $(addsuffix .,$(dir $(BLOB_H)))

$(BLOB_H): $(BLOB_HDR_DIR)/%.h: % $(BLOBS) | $(BLOB_DIRECTORIES)
```

Or if you want one to one mapping:

```
set_dir_dependency = $(foreach target,$(1),$(eval $1: | $(dir $(target)).))
$(call set_dir_dependency,$(BLOB_H))
```

IIRC, the `/.` syntax is to handle make <4 as currently used in OSx.
It should not be necessary anymore after 2020.01 as make <4 is deprecated in RIOT.


For the naming, I would not say it generates blobs as #11497 did, it is an `xxd` wrapper than only gives read only outputs.
From what I get this is only to replace:

https://github.com/RIOT-OS/RIOT/blob/255ce7d1a81ac41ffdc6732de7cab3dac76bde31/examples/javascript/Makefile#L58-L59



Adding directory variables creation is something that would be great globally  in Makefile.base and Makefile.include added on its own instead of added here in the middle of another things that does not really need it.

**Comment by kaspar030** [comment]
> It looks like the `SECONDEXPANSION` is only required to allow having subdirectories for headers inside `blob`. Is this a required feature?

It was asked for, and it makes sense, as e.g. jerryscript uses a "js/" subfolder.

**Comment by kaspar030** [comment]
> IIRC, the `/.` syntax is to handle make <4 as currently used in OSx.

correct.

File: makefiles/blob.inc.mk [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+#
+# Inspiration from:
+# http://ismail.badawi.io/blog/2017/03/28/automatic-directory-creation-in-make/
+#
+.PRECIOUS: $(BLOB_HDR_DIR)/. $(BLOB_HDR_DIR)%/.
+
+$(BLOB_HDR_DIR)/.:
+	@mkdir -p $@
+
+$(BLOB_HDR_DIR)%/.:
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply makefiles/blob.inc.mk]: It could, but as is it is self-contained.

File: makefiles/blob.inc.mk [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	@mkdir -p $@
+
+.SECONDEXPANSION:
+
+XXD ?= xxd
+
+$(BLOB_H): $(BLOB_HDR_DIR)/.
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply makefiles/blob.inc.mk]: not needed, I removed it.

File: makefiles/blob.inc.mk (hunk 7 above)
Comment (kaspar030) [reply makefiles/blob.inc.mk]: I've updated the comment.

Using `: |$(BLOB_H)`, no rebuild is triggered when changing a blob.

**Comment by kaspar030** [comment]
> Does it hurt?

@cladmi, other than issues with variables containing dollarsigns, do you know any downsides of enabling SECONDEXPANSION globally?

**Comment by kaspar030** [comment]
- split out the SECONDEXPANSION change (#12708), rebased on that.

**Comment by aabadie** [comment]
#12708, please rebase?

File: makefiles/application.inc.mk (hunk 4 above)
Comment (kaspar030) [reply makefiles/application.inc.mk]: I've added some doc to applications.inc.mk.

**Comment by kaspar030** [comment]
- rebased

File: tests/blob/Makefile [hunk 10]
```diff
@@ -0,0 +1,10 @@
+DEVELHELP ?= 0
```
Comment (aabadie) [inline tests/blob/Makefile]: Why disabling DEVELHELP by default? I would keep it.

File: tests/blob/tests/01-run.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    child.expect_exact("Hello blob_subdir!")
+    child.expect_exact("0x00")
+    child.expect_exact("0x01")
+    child.expect_exact("0x02")
+    child.expect_exact("0x03")
+    child.expect_exact("0xFF")
+
+
+if __name__ == "__main__":
+    sys.exit(run(testfunc, timeout=120))
```
Comment (aabadie) [inline tests/blob/tests/01-run.py]: I don't think the custom timeout is required. The test is running fast enough.

File: tests/blob/Makefile (hunk 10 above)
Comment (kaspar030) [reply tests/blob/Makefile]: yup, removed the line (default is 1)

File: tests/blob/tests/01-run.py (hunk 11 above)
Comment (kaspar030) [reply tests/blob/tests/01-run.py]: removed the timeout

**Comment by aabadie** [comment]
@bergzand, we need your approval here. Are your concerns addressed ?

**Comment by bergzand** [comment]
> @bergzand, we need your approval here. Are your concerns addressed ?

Yes

**Comment by kaspar030** [comment]
Thanks everyone!
